lib/pbench/server/database/models/audit.py

```diff
@@ -0,0 +1,370 @@
+
+    def __str__(self) -> str:
+        return f"Missing key value in {self.name!r}: {self.cause}"
+
+
+class AuditType(enum.Enum):
+    """
+    The type of Pbench Server resource affected by a CREATE, UPDATE, or DELETE
+    operation.
+    """
+
+    """Operation on a Dataset resource. This should include both the
+    object_id and object_name."""
+    DATASET = enum.auto()
+
+    """Operation on a system configuration setting. There's no meaningful id
+    or name; the 'attributes' column will record the actual attributes we
+    set."""
+    CONFIG = enum.auto()
+
+    """A default when no resource is affected"""
+    NONE = enum.auto()
+
+    """An Elasticsearch template. The base template name will be recorded in
+    object_name."""
+    TEMPLATE = enum.auto()
+
+    """Operation on an API token. There's no meaningful id or name."""
+    TOKEN = enum.auto()
+
+
+class AuditStatus(enum.Enum):
+
+    """Pending operation: this signals the beginning of an optation that might
+    fail or require further context later."""
+
+    BEGIN = enum.auto()
+
+    """Successful operation: an operation completed without problems."""
+    SUCCESS = enum.auto()
+
+    """Failed operation: an operation failed for reasons categorized by
+    AuditReason and possibly further detailed by a message."""
+    FAILURE = enum.auto()
+
+    """Warning: the operation didn't fail, but wasn't able to complete with
+    total success."""
+    WARNING = enum.auto()
+
+
+class AuditReason(enum.Enum):
+
+    """Permission denied."""
+
+    PERMISSION = enum.auto()
+
+    """Internal failure."""
+    INTERNAL = enum.auto()
+
+    """Consistency failure. For example, a corrupted tarball upload."""
+    CONSISTENCY = enum.auto()
+
+
+class Audit(Database.Base):
+    """
+    A framework to store Pbench audit records. These will track server
+    configuration changes as well as every mutation of user-visible data and
+    attribute each to a specific user and operation.
+
+    Architecturally, the SQL table may be used as a "front end" for a more
+    expensive audit repository (e.g., Elasticsearch). For example, we might
+    periodically bulk-replicate these rows as JSON data into an Elasticsearch
+    index and wipe the table.
+
+    Columns:
+        id: Generated unique ID of table row
+        root_id: Reference to the id of the operation's BEGIN audit record to
+            enable following an operation with multiple records.
+        name: Provide an optional name for the operation, like "upload" or
+            "index".
+        operation: The operation performed
+        object_type: The type of resource
+        object_id: The object on which an operation was performed
+        object_name: The name of the object (more meaningful than ID if the
+            resource has been deleted)
+        user_id: The identity of the user performing an operation
+        user_name: The name of the user (more meaningful than ID if the
+            user has been deleted)
```

Comment:
Perhaps we should audit the ID/Name pairs seen, since user IDs should be unique for all time, and then just have the audit store the ID only, and we can lookup the name when necessary.